core/descriptors/src/org/jetbrains/kotlin/name/NameUtils.kt

```diff
@@ -19,7 +19,7 @@ package org.jetbrains.kotlin.name
 import java.util.*
 
 object NameUtils {
-    private val SANITIZE_AS_JAVA_INVALID_CHARACTERS = "[^\\p{L}\\p{Digit}]".toRegex()
+    private val SANITIZE_AS_JAVA_INVALID_CHARACTERS = "[^\\p{L}\\p{Digit}\\$]".toRegex()
```

Comment:
Purpose of this change: LocalDeclarationsLowering uses sanitizeAsJavaIdentifier when naming the "lifted" local functions. The generated synthetic function for local function `bar` inside function `foo` is named `bar$default`. When lifted, the name should be `foo$bar$default` but without this change, `bar$default` is incorrectly sanitized and the lifted function name becomes `foo$bar_default`. This change ensures that `$` is not sanitized into `_`. ExpressionCodegen assumes that the names of the generated function end with `$default`.

Reply:
I'm not sure if this change is safe to make since this affects the current non-IR backend, but all the tests are passing. There doesn't seem to be any code within the compiler relying on `$` being sanitized into `_`, but it's possible there's some code out there that does.

Reply:
@punzki LocalNameProvider in LocalDeclarationsLowering is totally wrong concept. It was appeared only for fast bootstrapping. In some future it would be replaced with proper logic similar to one to CodegenAnnotatingVisitor in old backend. 
Seems name related problems will be solved with gettting rid of descriptors (https://upsource.jetbrains.com/kotlin/review/KOTLIN-CR-2697) and totally rewrited KotlinTypeMapper for IR.

Reply:
@max-kammerer Got it! Will hold off on this PR for now, and revisit it after that change is landed, and new type mapper is available. Is the type mapper in a different change under review, or still not in progress yet?

Reply:
I'm just starting it